diff --git a/common/os_arch_features.h b/common/os_arch_features.h
index ee0586d..75147ca 100644
--- a/common/os_arch_features.h
+++ b/common/os_arch_features.h
@@ -2,10 +2,10 @@
 
 #include <sys/types.h>
 #include <stddef.h>
+#include <sys/mman.h>
+#include <unistd.h>
 #include "pac_kit.h"
 
-#include "PlatformUnifiedInterface/platform.h"
-
 #if defined(__arm64e__) && __has_feature(ptrauth_calls)
 #include <ptrauth.h>
 #endif
@@ -36,10 +36,9 @@ template <typename T> inline T arm64e_pac_strip_and_sign(T &addr) {
 namespace android {
 inline void make_memory_readable(void *address, size_t size) {
 #if defined(ANDROID)
-  auto page = (void *)ALIGN_FLOOR(address, OSMemory::PageSize());
-  if (!OSMemory::SetPermission(page, OSMemory::PageSize(), kReadExecute)) {
-    return;
-  }
+  long page_size = sysconf(_SC_PAGESIZE);
+  auto page = (void *)ALIGN_FLOOR(address, page_size);
+  mprotect(page, page_size, PROT_READ | PROT_EXEC);
 #endif
 }
 } // namespace android
diff --git a/source/TrampolineBridge/ClosureTrampolineBridge/arm64/closure_bridge_arm64.asm b/source/TrampolineBridge/ClosureTrampolineBridge/arm64/closure_bridge_arm64.asm
index 7349181..59e7d4a 100644
--- a/source/TrampolineBridge/ClosureTrampolineBridge/arm64/closure_bridge_arm64.asm
+++ b/source/TrampolineBridge/ClosureTrampolineBridge/arm64/closure_bridge_arm64.asm
@@ -50,8 +50,7 @@ str TMP_REG_0, [sp, #(1 * 8)]
 // call convention: x0 = register context, x1 = interceptor entry
 mov x0, sp
 ldr x1, [sp, #(2 * 8 + 2 * 8 + 30 * 8 + 8 * 16)]
-adrp TMP_REG_0, cdecl(common_closure_bridge_handler)@PAGE
-add TMP_REG_0, TMP_REG_0, cdecl(common_closure_bridge_handler)@PAGEOFF
+ldr TMP_REG_0, common_closure_bridge_handler_addr
 blr TMP_REG_0
 
 // restore stack, saved original sp